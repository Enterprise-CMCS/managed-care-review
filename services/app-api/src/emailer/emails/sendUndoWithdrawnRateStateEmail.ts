import { formatCalendarDate } from '@mc-review/dates'
import type {
    ContractRevisionType,
    ProgramType,
    RateReviewType,
    RateType,
} from '../../domain-models'
import type { StateContact } from '../../gen/gqlServer'
import type { EmailData, EmailConfiguration } from '../emailer'
import { pruneDuplicateEmails } from '../formatters'
import { rateSummaryURL, submissionSummaryURL } from '../generateURLs'
import {
    findContractPrograms,
    renderTemplate,
    stripHTMLFromTemplate,
} from '../templateHelpers'
import { packageName as generatePackageName } from '@mc-review/hpp'

type ValidatedUnWithdrawnRate = {
    latestStatusAction: RateReviewType
    rateInfo: rateInfo
    stateContactEmails: string[]
    associatedContracts: associatedContracts[]
}

type associatedContracts = {
    contractName: string
    summaryURL: string
}

type undoWithdrawnRateEtaData = {
    rateInfo: rateInfo
    updatedBy: string
    updatedOn: string // mm/dd/yyyy format in PT timezone.
    reason: string
    associatedContracts: associatedContracts[]
}

type rateInfo = {
    rateName: string
    rateURL: string
}

//This parses for data related to associated contracts and contacts (used within validateAndParseUnwithdrawnRate)
const parseContractsAndContacts = (
    contractRevisions: ContractRevisionType[],
    statePrograms: ProgramType[],
    config: EmailConfiguration
) => {
    const stateContactEmails: string[] = []
    const associatedContracts: associatedContracts[] = []
    const seenContracts = new Set() //Used to avoid duplicates below

    //Iterate through all the contracts to save the contract name and generate a url - avoiding duplicates
    for (const contractRev of contractRevisions) {
        const pkgPrograms = findContractPrograms(contractRev, statePrograms)

        if (pkgPrograms instanceof Error) {
            return new Error(
                `Error parsing contract revision for package programs: ${contractRev.id}. ${pkgPrograms.message}`
            )
        }

        const contractInfo = {
            contractName: generatePackageName(
                contractRev.contract.stateCode,
                contractRev.contract.stateNumber,
                contractRev.formData.programIDs,
                pkgPrograms
            ),
            summaryURL: submissionSummaryURL(
                contractRev.contract.id,
                config.baseUrl
            ),
        }

        if (!seenContracts.has(contractInfo.contractName)) {
            seenContracts.add(contractInfo.contractName)
            associatedContracts.push(contractInfo)
        }

        contractRev.formData.stateContacts.forEach((contact: StateContact) => {
            if (contact.email) stateContactEmails.push(contact.email)
        })
    }

    return {
        associatedContracts,
        stateContactEmails,
    }
}

//Validates the unwithdrawn rate and parses it for necessary data
export const validateAndParseUnwithdrawnRate = (
    rate: RateType,
    statePrograms: ProgramType[],
    config: EmailConfiguration
): Error | ValidatedUnWithdrawnRate => {
    if (rate.consolidatedStatus !== 'RESUBMITTED') {
        return new Error('Rate consolidated status should be RESUBMITTED')
    }

    // Ensuring withdrawnFromContracts was cleared
    const withdrawnFromContracts = rate.withdrawnFromContracts
    if (withdrawnFromContracts && withdrawnFromContracts.length > 0) {
        return new Error('withdrawnFromContracts should be empty')
    }

    //check to make sure rate has actions
    const reviewStatusActions = rate.reviewStatusActions
    if (!reviewStatusActions || reviewStatusActions.length === 0) {
        return new Error('Rate does not have any review status actions')
    }

    //This gets us update info 'updatedBy/updatedAt/updateReason'
    const latestStatusAction = reviewStatusActions.sort(
        (a, b) =>
            new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
    )[0]
    if (latestStatusAction.actionType !== 'UNDER_REVIEW') {
        return new Error('Latest rate review action is not UNDER_REVIEW')
    }

    //This gets us the rate name and summary URL for the rate
    const rateName =
        rate.packageSubmissions[0].rateRevision.formData.rateCertificationName
    const rateID = rate.packageSubmissions[0].rateRevision.formData.rateID
    if (!rateName || !rateID) {
        return new Error(
            `Error parsing for ${rateID ? 'rateCertificationName' : 'rateID'}`
        )
    }

    const rateInfo: rateInfo = {
        rateName,
        rateURL: rateSummaryURL(rateID, config.baseUrl),
    }

    //Gathering contracts and state contacts associated to rate
    const contractRevisions = rate.packageSubmissions[0].contractRevisions
    if (contractRevisions.length === 0) {
        return new Error('Rate does not have any contract revisions')
    }
    const parsedContractsAndContacts = parseContractsAndContacts(
        contractRevisions,
        statePrograms,
        config
    )

    if (parsedContractsAndContacts instanceof Error) {
        return parsedContractsAndContacts
    }

    const { associatedContracts, stateContactEmails } =
        parsedContractsAndContacts

    return {
        latestStatusAction,
        rateInfo,
        associatedContracts,
        stateContactEmails,
    }
}

export const sendUndoWithdrawnRateStateEmail = async (
    rate: RateType,
    statePrograms: ProgramType[],
    config: EmailConfiguration
): Promise<EmailData | Error> => {
    const undoWithdrawnRateData = validateAndParseUnwithdrawnRate(
        rate,
        statePrograms,
        config
    )

    if (undoWithdrawnRateData instanceof Error) {
        return undoWithdrawnRateData
    }

    const {
        latestStatusAction, //Contains update information
        rateInfo, //Rate name and summary URL
        associatedContracts, //Contracts connected to the rate
        stateContactEmails,
    } = undoWithdrawnRateData

    const toAddresses = pruneDuplicateEmails([
        ...stateContactEmails,
        ...config.devReviewTeamEmails,
    ])

    const etaData: undoWithdrawnRateEtaData = {
        rateInfo: rateInfo,
        updatedBy: latestStatusAction.updatedBy.email,
        updatedOn: formatCalendarDate(
            latestStatusAction.updatedAt,
            'America/Los_Angeles'
        ),
        reason: latestStatusAction.updatedReason,
        associatedContracts: associatedContracts,
    }

    const template = await renderTemplate<undoWithdrawnRateEtaData>(
        'sendUndoWithdrawnRateStateEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.rateInfo.rateName} was unwithdrawn`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
