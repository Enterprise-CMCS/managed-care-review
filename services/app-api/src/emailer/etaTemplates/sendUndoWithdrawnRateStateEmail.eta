<h1 style=<%~'"font-weight: normal;"'%>>Rate <%= it.rateName %> status was updated to 'Submitted' by CMS</h1>
<b>Updated by:</b> <%= it.updatedBy %><br />
<b>Updated on:</b> <%= it.updatedOn %><br />
<b>Reason:</b> <%= it.reason %><br />
<b>Related submission:</b>
<% if (it.associatedContracts.length > 1) { %>
<ul>
<% it.associatedContracts.forEach(function(contract){ %>
  <li><a href="<%= it.summaryURL %>"><%= contract.contractName %></a></li>
<% }) %>
</ul>
<% } else { %>
<a href="<%= it.summaryURL %>"><%= it.associatedContracts[0].contractName %></a><br />
<br />
<% } %>