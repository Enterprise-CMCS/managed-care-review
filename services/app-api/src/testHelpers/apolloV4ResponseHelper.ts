import type { GraphQLFormattedError } from 'graphql'
import type { ApolloServer } from '@apollo/server'

export type TestGraphQLResponse<T = any> = {
    data?: T
    errors?: GraphQLFormattedError[]
}

export function extractGraphQLResponse<T = any>(
    response: any
): TestGraphQLResponse<T> {
    if ('body' in response && response.body) {
        const result =
            response.body.kind === 'single'
                ? response.body.singleResult
                : response.body

        return {
            data: result.data,
            errors: result.errors,
        }
    }

    return {
        data: response.data,
        errors: response.errors,
    }
}

export async function executeGraphQLOperation<TData = any, TVariables = any>(
    server: ApolloServer,
    options: {
        query: any
        variables?: TVariables
        contextValue?: any
    }
): Promise<TestGraphQLResponse<TData>> {
    const response = await server.executeOperation({
        query: options.query,
        variables: options.variables as any,
    })

    return extractGraphQLResponse<TData>(response)
}
