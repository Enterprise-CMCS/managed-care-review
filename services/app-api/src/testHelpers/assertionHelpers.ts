// Functions that assists with type narrow in jest tests

// must -  interrupts test flow to through an error
function must<T>(maybeErr: T | Error): T {
    if (maybeErr instanceof Error) {
        throw maybeErr
    }
    return maybeErr
}

// expectToBeDefined - properly  type narrows the results of toBeDefined
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41179 would eventually address this but is not implemented
function expectToBeDefined<T>(arg: T): asserts arg is NonNullable<T> {
    expect(arg).toBeDefined()
}

export { must, expectToBeDefined }
