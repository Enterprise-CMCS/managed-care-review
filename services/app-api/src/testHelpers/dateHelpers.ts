// returns today's date in the ISO standard form. like '2022-02-24'
function todaysDate(): string {
    const today = new Date()
    return today.toISOString().split('T')[0]
}

export { todaysDate }
