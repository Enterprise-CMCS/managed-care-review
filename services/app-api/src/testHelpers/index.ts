export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './assertionHelpers'

export {
    mockInsertContractArgs,
    mockContractRevision,
} from './contractDataMocks'

export {
    mockInsertRateArgs,
    mockRateRevision,
    mockDraftRate,
} from './rateDataMocks'

export { testS3Client } from './s3Helpers'
export { getStateRecord } from './stateHelpers'

export { consoleLogFullData } from './debugHelpers'

export {
    fetchTestRateById,
    submitTestRate,
    unlockTestRate,
    updateTestRate,
} from './gqlRateHelpers'

export {
    createAndSubmitTestContract,
    fetchTestContract,
    approveTestContract,
    updateTestContractDraftRevision,
    createTestContract,
    fetchTestContractWithQuestions,
} from './gqlContractHelpers'

export {
    clearDocMetadata,
    clearMetadataFromContractFormData,
    clearMetadataFromRateFormData,
} from './documentHelpers'

export { mockGqlContractDraftRevisionFormDataInput } from './gqlContractInputMocks'
