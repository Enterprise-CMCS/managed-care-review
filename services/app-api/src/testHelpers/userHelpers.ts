import { NewPostgresStore, type InsertUserArgsType } from '../postgres'
import type {
    AdminUserType,
    CMSApproverUserType,
    CMSUserType,
    StateUserType,
    UserType,
    BusinessOwnerUserType,
    HelpdeskUserType,
} from '../domain-models'
import { sharedTestPrismaClient } from './storeHelpers'
import { v4 as uuidv4 } from 'uuid'

const testCMSUser = (userData?: Partial<CMSUserType>): CMSUserType => ({
    id: uuidv4(),
    role: 'CMS_USER',
    email: 'zuko@example.com',
    familyName: 'Zuko',
    givenName: 'Prince',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testCMSApproverUser = (
    userData?: Partial<CMSApproverUserType>
): CMSApproverUserType => ({
    id: uuidv4(),
    role: 'CMS_APPROVER_USER',
    email: 'azula@example.com',
    familyName: 'Azula',
    givenName: 'Princess',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testStateUser = (userData?: Partial<StateUserType>): StateUserType => ({
    id: uuidv4(),
    stateCode: 'FL',
    role: 'STATE_USER',
    email: 'james@example.com',
    familyName: 'Brown',
    givenName: 'James',
    ...userData,
})

const testAdminUser = (userData?: Partial<AdminUserType>): AdminUserType => ({
    id: uuidv4(),
    role: 'ADMIN_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testBusinessOwnerUser = (
    userData?: Partial<BusinessOwnerUserType>
): BusinessOwnerUserType => ({
    id: uuidv4(),
    role: 'BUSINESSOWNER_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const testHelpdeskUser = (
    userData?: Partial<HelpdeskUserType>
): HelpdeskUserType => ({
    id: uuidv4(),
    role: 'HELPDESK_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const createDBUsersWithFullData = async (
    users: UserType[]
): Promise<UserType[]> => {
    const prismaClient = await sharedTestPrismaClient()
    const postgresStore = NewPostgresStore(prismaClient)

    const usersSeed: InsertUserArgsType[] = users.map((user: UserType) => {
        const userArgs: InsertUserArgsType & { id?: string } = {
            ...user,
            userID: user.id,
        }

        delete userArgs.id

        return userArgs as InsertUserArgsType
    })

    const result = await postgresStore.insertManyUsers(usersSeed)

    if (result instanceof Error) {
        throw result
    }

    return result
}

const iterableCmsUsersMockData: {
    userRole: 'CMS_USER' | 'CMS_APPROVER_USER'
    mockUser: <T>(userData?: Partial<T>) => CMSUserType | CMSApproverUserType
}[] = [
    {
        userRole: 'CMS_USER',
        mockUser: testCMSUser,
    },
    {
        userRole: 'CMS_APPROVER_USER',
        mockUser: testCMSApproverUser,
    },
]

const iterableNonCMSUsersMockData: {
    userRole:
        | 'HELPDESK_USER'
        | 'BUSINESSOWNER_USER'
        | 'ADMIN_USER'
        | 'STATE_USER'
    mockUser: <T>(
        userData?: Partial<T>
    ) =>
        | AdminUserType
        | BusinessOwnerUserType
        | HelpdeskUserType
        | StateUserType
}[] = [
    {
        userRole: 'ADMIN_USER',
        mockUser: testAdminUser,
    },
    {
        userRole: 'BUSINESSOWNER_USER',
        mockUser: testBusinessOwnerUser,
    },
    {
        userRole: 'HELPDESK_USER',
        mockUser: testHelpdeskUser,
    },
    {
        userRole: 'STATE_USER',
        mockUser: testStateUser,
    },
]

export {
    testAdminUser,
    testStateUser,
    testCMSUser,
    createDBUsersWithFullData,
    iterableCmsUsersMockData,
    iterableNonCMSUsersMockData,
    testCMSApproverUser,
    testBusinessOwnerUser,
    testHelpdeskUser,
}
