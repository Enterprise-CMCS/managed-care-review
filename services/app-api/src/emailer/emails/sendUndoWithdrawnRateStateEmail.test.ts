import { testCMSUser, testStateUser } from '../../testHelpers/userHelpers'
import {
    mockContract,
    mockContractRev,
    mockMNState,
    mockRate,
    testEmailConfig,
} from '../../testHelpers/emailerHelpers'
import { sendUndoWithdrawnRateStateEmail } from './sendUndoWithdrawnRateStateEmail'

const testRate = mockRate({
    id: 'test-rate',
    createdAt: new Date('2025-03-05'),
    updatedAt: new Date('2025-03-05'),
    status: 'RESUBMITTED',
    reviewStatus: 'UNDER_REVIEW',
    consolidatedStatus: 'RESUBMITTED',
    parentContractID: 'parent-contract',
    withdrawnFromContracts: [],
    packageSubmissions: [
        {
            submitInfo: {
                updatedAt: new Date('2025-03-05'),
                updatedReason: 'Messed up - undoing withdraw',
                updatedBy: testStateUser({
                    email: 'undoer-of-withdraws@example.com',
                }),
            },
            submittedRevisions: [],
            rateRevision: {
                id: 'test-rate-revision',
                rateID: 'test-rate',
                submitInfo: {
                    updatedAt: new Date('2025-03-05'),
                    updatedReason: 'Messed up - undoing withdraw',
                    updatedBy: testCMSUser({
                        email: 'undoer-of-withdraws@example.com',
                    }),
                },
                createdAt: new Date('2025-03-05'),
                updatedAt: new Date('2025-03-05'),
                formData: {
                    rateType: 'NEW',
                    rateCapitationType: 'RATE_CELL',
                    rateDateStart: new Date('2025-03-05'),
                    rateDateEnd: new Date('2025-03-05'),
                    rateDateCertified: new Date('2025-03-05'),
                    rateCertificationName: 'test-rate-certification-name',
                    rateID: 'test-rate',
                    rateProgramIDs: [mockMNState().programs[0].id],
                    rateDocuments: [
                        {
                            s3URL: 's3://bucketname/key/test1',
                            name: 'ratedoc1.doc',
                            sha256: 'foobar',
                        },
                    ],
                    supportingDocuments: [],
                    certifyingActuaryContacts: [
                        {
                            name: 'Foo Person',
                            titleRole: 'Bar Job',
                            email: 'certifyingActuary@example.com',
                            actuarialFirm: 'GUIDEHOUSE',
                        },
                    ],
                    addtlActuaryContacts: [
                        {
                            name: 'Bar Person',
                            titleRole: 'Baz Job',
                            email: 'addtlActuaryContacts@example.com',
                            actuarialFirm: 'OTHER',
                            actuarialFirmOther: 'Some Firm',
                        },
                    ],
                    actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                },
            },
            contractRevisions: [
                mockContractRev({
                    submitInfo: {
                        updatedAt: new Date('2025-03-05'),
                        updatedReason: 'Messed up - undoing withdraw',
                        updatedBy: testCMSUser({
                            email: 'undoer-of-withdraws@example.com',
                        }),
                    },
                    formData: {
                        ...mockContractRev().formData,
                        stateContacts: [
                            {
                                name: 'parent-contract-state-contact',
                                titleRole: 'state contact',
                                email: 'parent-contract-state-contact@state.com',
                            },
                        ],
                    },
                }),
            ],
        },
    ],
    reviewStatusActions: [
        {
            rateID: 'test-rate',
            updatedBy: testCMSUser(),
            updatedReason: 'I am undoing a withdraw',
            actionType: 'UNDER_REVIEW',
            updatedAt: new Date('2025-03-05'),
        },
    ],
})

const statePrograms = mockMNState().programs

describe('sendUndoWithdrawnRateStateEmail', () => {
    it('renders email for unwithdrawn rate as expected', async () => {
        const template = await sendUndoWithdrawnRateStateEmail(
            testRate,
            statePrograms,
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })

    it('includes expected state contacts in toAddress', async () => {
        const template = await sendUndoWithdrawnRateStateEmail(
            testRate,
            statePrograms,
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                'parent-contract-state-contact@state.com',
                'devreview1@example.com',
                'devreview2@example.com',
            ])
        )
    })
})

describe('sendUndoWithdrawnRateStateEmail error handling', () => {
    it('returns an error if rate consolidatedStatus is not RESUBMITTED', async () => {
        const rateWithWrongStatus = {
            ...testRate,
            consolidatedStatus: 'WITHDRAWN' as const,
        }
        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithWrongStatus,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate consolidated status should be RESUBMITTED'
        )
    })

    it('returns an error if the withdrawnFromContracts array was not emptied', async () => {
        const rateWithWrongStatus = {
            ...testRate,
            withdrawnFromContracts: [
                {
                    ...mockContract({
                        id: 'parent-contract',
                        stateNumber: 100,
                        packageSubmissions: [
                            {
                                submitInfo: {
                                    updatedAt: new Date('2025-03-05'),
                                    updatedReason:
                                        'Messed up - undoing withdraw',
                                    updatedBy: testCMSUser({
                                        email: 'undoer-of-withdraws@example.com',
                                    }),
                                },
                                submittedRevisions: [],
                                rateRevisions: [],
                                contractRevision: mockContractRev({
                                    submitInfo: {
                                        updatedAt: new Date('2025-03-05'),
                                        updatedReason:
                                            'Messed up - undoing withdraw',
                                        updatedBy: testCMSUser({
                                            email: 'undoer-of-withdraws@example.com',
                                        }),
                                    },
                                }),
                            },
                        ],
                    }),
                },
            ],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithWrongStatus,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'withdrawnFromContracts should be empty'
        )
    })

    it('returns an error if there are no review status actions present', async () => {
        const missingReviewStatusActions = {
            ...testRate,
            reviewStatusActions: [],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            missingReviewStatusActions,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate does not have any review status actions'
        )
    })

    it('returns an error if the latest review action is NOT "UNDER_REVIEW"', async () => {
        const rateWithWrongActionType = {
            ...testRate,
            reviewStatusActions: [
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'I must withdraw',
                    actionType: 'WITHDRAW' as const,
                    updatedAt: new Date('2025-03-10'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test under review',
                    actionType: 'UNDER_REVIEW' as const,
                    updatedAt: new Date('2025-03-05'),
                },
                {
                    rateID: 'test-rate',
                    updatedBy: testCMSUser(),
                    updatedReason: 'Test withdraw',
                    actionType: 'WITHDRAW' as const,
                    updatedAt: new Date('2025-03-15'),
                },
            ],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithWrongActionType,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest rate review action is not UNDER_REVIEW'
        )
    })

    it('returns an error if there is no rateCertificationName', async () => {
        const rateWithoutRateCertificationName = {
            ...testRate,
            packageSubmissions: [
                {
                    rateRevision: {
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName: '',
                            rateID: 'test-rate',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                    contractRevisions: [],
                },
            ],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithoutRateCertificationName,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Error parsing for rateCertificationName'
        )
    })

    it('returns an error if there is no rateID', async () => {
        const rateWithoutRateCertificationName = {
            ...testRate,
            packageSubmissions: [
                {
                    rateRevision: {
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName: 'test-rate-name',
                            rateID: '',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                    contractRevisions: [],
                },
            ],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithoutRateCertificationName,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe('Error parsing for rateID')
    })

    it('returns an error if there are no contractRevisions', async () => {
        const rateWithoutContractRevisions = {
            ...testRate,
            packageSubmissions: [
                {
                    rateRevision: {
                        id: 'test-rate-revision',
                        rateID: 'test-rate',
                        submitInfo: {
                            updatedAt: new Date('2025-03-05'),
                            updatedReason: 'Messed up - undoing withdraw',
                            updatedBy: testCMSUser({
                                email: 'undoer-of-withdraws@example.com',
                            }),
                        },
                        createdAt: new Date('2025-03-05'),
                        updatedAt: new Date('2025-03-05'),
                        formData: {
                            rateType: 'NEW',
                            rateCapitationType: 'RATE_CELL',
                            rateDateStart: new Date('2025-03-05'),
                            rateDateEnd: new Date('2025-03-05'),
                            rateDateCertified: new Date('2025-03-05'),
                            rateCertificationName:
                                'test-rate-certification-name',
                            rateID: 'test-rate',
                            rateProgramIDs: [mockMNState().programs[0].id],
                            rateDocuments: [
                                {
                                    s3URL: 's3://bucketname/key/test1',
                                    name: 'ratedoc1.doc',
                                    sha256: 'foobar',
                                },
                            ],
                            supportingDocuments: [],
                            certifyingActuaryContacts: [
                                {
                                    name: 'Foo Person',
                                    titleRole: 'Bar Job',
                                    email: 'certifyingActuary@example.com',
                                    actuarialFirm: 'GUIDEHOUSE',
                                },
                            ],
                            addtlActuaryContacts: [
                                {
                                    name: 'Bar Person',
                                    titleRole: 'Baz Job',
                                    email: 'addtlActuaryContacts@example.com',
                                    actuarialFirm: 'OTHER',
                                    actuarialFirmOther: 'Some Firm',
                                },
                            ],
                            actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                        },
                    },
                    contractRevisions: [],
                },
            ],
        }

        const template = await sendUndoWithdrawnRateStateEmail(
            rateWithoutContractRevisions,
            statePrograms,
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Rate does not have any contract revisions'
        )
    })
})
