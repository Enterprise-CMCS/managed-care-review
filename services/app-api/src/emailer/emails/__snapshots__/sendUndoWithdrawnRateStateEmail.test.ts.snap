// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendUndoWithdrawnRateStateEmail > renders email for unwithdrawn rate as expected 1`] = `
"<h1 style="font-weight: normal;">Rate test-rate-certification-name status was updated to 'Submitted' by CMS</h1>
<b>Updated by:</b> zuko@example.com<br />
<b>Updated on:</b> 03/04/2025<br />
<b>Reason:</b> I am undoing a withdraw<br />
<b>Related submission:</b>
<a href="http://localhost/submissions/12345">MCR-MN-0003-SNBC</a><br />
<br />
If you have any questions, please reach out to your CMS point of contact.
<br />
  <a href="http://localhost/submissions/12345">View the submission in MC-Review</a>
"
`;
